game/scripts/vscripts/Goals.ts

```diff
@@ -0,0 +1,123 @@
+export abstract class TrackableGoal {
+    private _completed = false
+    private _started = false
+
+    /**
+     * Marks the goal as completed. Also marks it as started if it wasn't started already.
+     */
+    complete() {
+        this._started = true
+        this._completed = true
+    }
+
+    /**
+     * Marks the goal as started.
+     */
+    start() {
+        this._started = true
+    }
+
+    /**
+     * Whether the goal was completed.
+     */
+    get completed() { return this._completed }
+
+    /**
+     * Whether the goal was started.
+     */
+    get started() { return this._started }
+
+    abstract getGoal(): Goal
+}
+
+export class TrackableGoalBoolean extends TrackableGoal {
+    private _value: number = 0
+
+    constructor(readonly text: string) {
+        super()
+    }
+
+    getGoal(): Goal {
+        return {
+            completed: this.completed,
+            text: this.text
+        }
+    }
+}
+
+export class TrackableGoalNumeric extends TrackableGoal {
+    private _value: number = 0
+
+    constructor(readonly text: string, readonly maximum: number) {
+        super()
+    }
+
+    /**
+     * Sets the value of the numeric goal.
+     * @param value Value of the numeric goal.
+     */
+    setValue(value: number) {
+        if (value < 0 || value > this.maximum) {
+            Warning("Set numeric goal value outside of its range")
+        }
+
+        this._value = value
+    }
+
+    /**
+     * Gets the value of the numeric goal.
+     */
+    get value() {
+        return
+    }
+
+    getGoal(): Goal {
+        return {
+            completed: this.completed,
+            text: this.text + ": " + this._value + "/" + this.maximum
```

Comment:
This would make the text non-localizable I guess so we'd need to implement this in panorama?

Reply:
New idea is to implement a function in panorama that replaces stuff in eg. dollar signs by its localization. Then this would be easy to do and could be kept almost as is.

Reply:
Another method we talked about before: when sending panorama the event for setting a Goal, optionally also send current value and maximum values. Panorama then does the concatenation after localization.

Reply:
Just for the record: I think Shush's solution here is the way to go.